source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,793 @@
+  LoopUnrollState(ir::Instruction* induction, ir::BasicBlock* continue_block,
+                  ir::BasicBlock* condition)
+      : previous_phi_(induction),
+        previous_continue_block_(continue_block),
+        previous_condition_block_(condition),
+        new_phi(nullptr),
+        new_continue_block(nullptr),
+        new_condition_block(nullptr),
+        new_header_block(nullptr) {}
+
+  // Swap the state so that the new nodes are now the previous nodes.
+  void NextIterationState() {
+    previous_phi_ = new_phi;
+    previous_continue_block_ = new_continue_block;
+    previous_condition_block_ = new_condition_block;
+
+    // Clear new nodes.
+    new_phi = nullptr;
+    new_continue_block = nullptr;
+    new_condition_block = nullptr;
+    new_header_block = nullptr;
+
+    // Clear new block/instruction maps.
+    new_blocks.clear();
+    new_inst.clear();
+  }
+
+  // The induction variable from the immediately preceding loop body.
+  ir::Instruction* previous_phi_;
+
+  // The previous continue block. The backedge will be removed from this and
+  // added to the new continue block.
+  ir::BasicBlock* previous_continue_block_;
+
+  // The previous condition block. This may be folded to flatten the loop.
+  ir::BasicBlock* previous_condition_block_;
+
+  // The new induction variable.
+  ir::Instruction* new_phi;
+
+  // The new continue block.
+  ir::BasicBlock* new_continue_block;
+
+  // The new condition block.
+  ir::BasicBlock* new_condition_block;
+
+  // The new header block.
+  ir::BasicBlock* new_header_block;
+
+  // A mapping of new block ids to the original blocks which they were copied
+  // from.
+  std::unordered_map<uint32_t, ir::BasicBlock*> new_blocks;
+
+  // A mapping of new instruction ids to the instruction ids from which they
+  // were copied.
+  std::unordered_map<uint32_t, uint32_t> new_inst;
+};
+
+// This class implements the actual unrolling. It uses a LoopUnrollState to
+// maintain the state of the unrolling inbetween steps.
+class LoopUnrollerUtilsImpl {
+ public:
+  using BasicBlockListTy = std::vector<std::unique_ptr<ir::BasicBlock>>;
+
+  LoopUnrollerUtilsImpl(ir::IRContext* c, ir::Function* function)
+      : ir_context_(c),
+        function_(*function),
+        loop_condition_block_(nullptr),
+        loop_induction_variable_(nullptr),
+        number_of_loop_iterations_(0) {}
+
+  // Unroll the |loop| by given |factor| by copying the whole body |factor|
+  // times. The resulting basicblock structure will remain a loop.
+  void PartiallyUnroll(ir::Loop*, size_t factor);
+
+  // If partially unrolling the |loop| would leave the loop with too many bodies
+  // for its number of iterations then this method should be used. This method
+  // will duplicate the |loop| compltetely, makingthe duplicated loop the
+  // successor of the originals merge block. The original loop will have its
+  // condition changed to loop over the uneven part and the duplicate will be
```

Comment:
A better name than "uneven part" might be the "residual".  I think this is commonly used in the compiler literature.  Just something to consider.  If you are following the language defined in a particular paper, add a comment at the top that you are following the implementation in that paper.